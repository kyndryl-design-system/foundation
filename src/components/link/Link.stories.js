import { html } from 'lit';
import './link';

export default {
  title: 'Components/Link',
  component: 'kd-link',
  argTypes: {},
};

const args = {
  href: '',
};

export const Link = {
  args: args,
  render: (args) => html` <kd-link href=${args.href}> Link1 </kd-link> `,
};
