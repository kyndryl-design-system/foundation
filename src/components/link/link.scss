@use '../../scss/global.scss';

:host {
  display: inline-block;
  margin: 4px 0;
}

// base class
.kd-link-text {
  display: inline-flex;
  padding: 0 2px;
  position: relative;
  text-decoration: none;
  -webkit-transition: color 300ms ease-in-out;
  -moz-transition: color 300ms ease-in-out;
  -o-transition: color 300ms ease-in-out;
  transition: color 300ms ease-in-out;

  &:hover {
    ::slotted(kd-icon) {
      transform: translate(4px, 0);
    }
  }

  // Primary (app)
  &-primary {
    color: var(--kd-color-text-link);

    &:hover {
      color: var(--kd-color-text-link-hover);
    }
    &:focus {
      color: var(--kd-color-text-tertiary);
      outline: 1px solid var(--kd-color-border-focus);
      -webkit-transition: outline 100ms ease-in-out;
      -moz-transition: outline 100ms ease-in-out;
      -o-transition: outline 100ms ease-in-out;
      transition: outline 100ms ease-in-out;
      text-decoration: none;
    }
    &:active {
      color: var(--kd-color-text-link-pressed);
      text-decoration: underline;
      outline: none;
    }
  }

  // Secondary (web)
  &-secondary {
    color: var(--kd-color-text-primary);

    &:hover {
      color: var(--kd-color-text-primary-hover);
    }
    &:focus {
      outline: 1px solid var(--kd-color-border-focus);
      -webkit-transition: outline 100ms ease-in-out;
      -moz-transition: outline 100ms ease-in-out;
      -o-transition: outline 100ms ease-in-out;
      transition: outline 100ms ease-in-out;
      text-decoration: none;
    }
    &:active {
      color: var(--kd-color-text-primary-pressed);
      text-decoration: underline;
      outline: none;
    }
  }
  // Link disable class
  &-disabled {
    cursor: not-allowed;
    color: var(--kd-color-darkstone-40);
    text-decoration: none;
    border: none;
    outline: none;
    pointer-events: none;

    &:hover,
    &:focus,
    &:active,
    &:visited {
      color: var(--kd-color-darkstone-40);
      border: none;
      outline: none;
      text-decoration: none;
      slot[name='icon']::slotted(*) {
        transform: translate(0px, 0);
      }
    }
  }

  // Link Inline - standalone classes
  &-inline {
    text-decoration: underline;
  }
  &-standalone {
    text-decoration: none;
    &:hover {
      text-decoration: underline;
    }
  }
  &-span-flex {
    display: flex;
    align-items: center;

    slot[name='icon']::slotted(*) {
      display: inline-flex;
      align-self: center;
      margin-left: 0px;
      transition: 300ms ease-in-out;
      -moz-transition: 300ms ease-in-out;
      -o-transition: 300ms ease-in-out;
      -webkit-transition: 300ms ease-in-out;
    }
  }
}
