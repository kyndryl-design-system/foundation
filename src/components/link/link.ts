import { LitElement, html } from 'lit';
import { customElement, property } from 'lit/decorators.js';
import { classMap } from 'lit-html/directives/class-map.js';
import { LINK_TYPES, LINK_TARGETS } from './defs';

import LinkStyles from './link.scss';

/**
 * Component for navigation links.
 * @fires on-click - Captures the click event and emits the original event details.
 * @slot unnamed - Slot for link text.
 * @slot icon - Slot for an icon.
 */

@customElement('kd-link')
export class Link extends LitElement {
  static override styles = [LinkStyles];

  /** Link url. */
  @property({ type: String })
  href = '';

  /** Defines a target attribute for where to load the URL. Possible options include "_self" (deafult), "_blank", "_parent", "_top" */
  @property({ type: String })
  target: LINK_TARGETS = LINK_TARGETS.SELF;

  /** The Link type. Primary(App) or Secondary(Web).*/
  @property({ type: String })
  kind: LINK_TYPES = LINK_TYPES.PRIMARY;

  /** Defines a relationship between a linked resource and the document. An empty string (default) means no particular relationship */
  @property({ type: String })
  rel = '';

  /** Determines if the link is disabled.*/
  // Reference for disabled links:
  // https://www.scottohara.me/blog/2021/05/28/disabled-links.html
  @property({ type: Boolean, reflect: true })
  disabled = false;

  /** Whether you want the standalone Link. Bydefault false. Use this prop. (true) with icon with link variant. */
  @property({ type: Boolean })
  standalone = false;

  override render() {
    const classes = this.returnClassMap();

    return html`<a
      class="kd-link-text ${classes}"
      target=${this.target}
      role="link"
      rel=${this.rel}
      href=${this.href ? this.href : 'javascript:void(0)'}
      ?disabled=${this.disabled}
      @click=${(e: Event) => this.handleClick(e)}
    >
      <span class="kd-link-text-span-flex">
        <slot></slot>
        <slot name="icon"></slot>
      </span>
    </a>`;
  }
  // -- Apply classes according to states, kind etc. -- //
  private returnClassMap() {
    if (this.disabled) {
      return classMap({
        ['kd-link-text-disabled']: this.disabled,
      });
    } else {
      return classMap({
        ['kd-link-text-primary']:
          this.kind === LINK_TYPES.PRIMARY || !this.kind,
        ['kd-link-text-secondary']: this.kind === LINK_TYPES.SECONDARY,
        ['kd-link-text-inline']: !this.standalone,
        ['kd-link-text-standalone']: this.standalone,
      });
    }
  }

  private handleClick(e: Event) {
    if (this.disabled) {
      e.preventDefault();
      return;
    }
    const event = new CustomEvent('on-click', {
      detail: { origEvent: e },
    });
    this.dispatchEvent(event);
  }
}
declare global {
  interface HTMLElementTagNameMap {
    'kd-link': Link;
  }
}
