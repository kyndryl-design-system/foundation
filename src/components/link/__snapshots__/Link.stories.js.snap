// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/Link InteractionTests play-test 1`] = `
<kd-link id="test"
         href
         target="_self"
         kind="primary"
>
  Link Text
</kd-link>
<a role="link"
   class="kd-link-text  kd-link-text-primary kd-link-text-inline "
   target="_self"
   rel
   href="javascript:void(0)"
>
  <span class="kd-link-text-span-flex">
    <slot>
    </slot>
    <slot name="icon">
    </slot>
  </span>
</a>
`;

exports[`Components/Link Link smoke-test 1`] = `
<kd-link id="test"
         href
         target="_self"
         kind="primary"
>
  Link Text
</kd-link>
<a role="link"
   class="kd-link-text  kd-link-text-primary kd-link-text-inline "
   target="_self"
   rel
   href="javascript:void(0)"
>
  <span class="kd-link-text-span-flex">
    <slot>
    </slot>
    <slot name="icon">
    </slot>
  </span>
</a>
`;

exports[`Components/Link LinkWithIcon smoke-test 1`] = `
<kd-link id="test"
         standalone
         href
         target="_self"
         kind="primary"
>
  Link Text
  <kd-icon slot="icon"
           sizeoverride="16"
  >
  </kd-icon>
</kd-link>
<a role="link"
   class="kd-link-text  kd-link-text-primary kd-link-text-standalone "
   target="_self"
   rel
   href="javascript:void(0)"
>
  <span class="kd-link-text-span-flex">
    <slot>
    </slot>
    <slot name="icon">
    </slot>
  </span>
</a>
`;
